eth-rpc/server/servertypes.nim

```diff
@@ -13,15 +14,152 @@ type
+
+proc unRegisterAll*(server: RpcServer) = server.procs.clear
+
+proc newRpcServer*(address = "localhost", port: Port = Port(8545)): RpcServer =
+  result = RpcServer(
     socket: newAsyncSocket(),
     port: port,
     address: address,
     procs: newTable[string, RpcProc]()
   )
 
-proc register*(server: RpcServer, name: string, rpc: RpcProc) =
-  server.procs[name] = rpc
+var sharedServer: RpcServer
 
-proc unRegisterAll*(server: RpcServer) = server.procs.clear
+proc sharedRpcServer*(): RpcServer =
+  if sharedServer.isNil: sharedServer = newRpcServer("")
+  result = sharedServer
+  
+proc fromJson(n: JsonNode, argName: string, result: var bool) =
+  if n.kind != JBool: raise newException(ValueError, "Parameter \"" & argName & "\" expected JBool but got " & $n.kind)
+  result = n.getBool()
+
+proc fromJson(n: JsonNode, argName: string, result: var int) =
+  if n.kind != JInt: raise newException(ValueError, "Parameter \"" & argName & "\" expected JInt but got " & $n.kind)
+  result = n.getInt()
+
+proc fromJson(n: JsonNode, argName: string, result: var byte) =
+  if n.kind != JInt: raise newException(ValueError, "Parameter \"" & argName & "\" expected JInt but got " & $n.kind)
+  let v = n.getInt()
+  if v > 255 or v < 0: raise newException(ValueError, "Parameter \"" & argName & "\" value out of range for byte: " & $v)
+  result = byte(v)
+
+proc fromJson(n: JsonNode, argName: string, result: var float) =
+  if n.kind != JFloat: raise newException(ValueError, "Parameter \"" & argName & "\" expected JFloat but got " & $n.kind)
+  result = n.getFloat()
+
+proc fromJson(n: JsonNode, argName: string, result: var string) =
+  if n.kind != JString: raise newException(ValueError, "Parameter \"" & argName & "\" expected JString but got " & $n.kind)
+  result = n.getStr()
+
+proc fromJson[T](n: JsonNode, argName: string, result: var seq[T]) =
+  result = newSeq[T](n.len)
+  for i in 0 ..< n.len:
+    fromJson(n[i], argName, result[i])
+
+proc fromJson[N, T](n: JsonNode, argName: string, result: var array[N, T]) =
+  for i in 0 ..< n.len:
+    fromJson(n[i], argName, result[i])
+
+proc fromJson[T: object](n: JsonNode, argName: string, result: var T) =
+  for k, v in fieldpairs(result):
+    fromJson(n[k], k, v)
+
+proc unpackArg[T](argIdx: int, argName: string, argtype: typedesc[T], args: JsonNode): T =
+  when argType is array or argType is seq:
+    if args[argIdx].kind != JArray: raise newException(ValueError, "Parameter \"" & argName & "\" expected JArray but got " & $args[argIdx].kind)
+  when argType is array:
+    if args[argIdx].len > result.len: raise newException(ValueError, "Parameter \"" & argName & "\" item count is too big for array")
+  when argType is object:
+    if args[argIdx].kind != JObject: raise newException(ValueError, "Parameter \"" & argName & "\" expected JObject but got " & $args[argIdx].kind)
+  fromJson(args[argIdx], argName, result)
+
+proc expectArrayLen(node: NimNode, paramsIdent: untyped, length: int) =
```

Comment:
Sorry, what i meant by this is not a compile-time proc, but a simple non-generic runtime proc, and the `setupParams` would just add a call to it, if that makes sense. Sorry for the confusion. Anyways, this is by no means critical.

Reply:
@yglukhov I've replaced the macro now :)